### ISSUE TITLE
sti scripts - purpose, usage, names

### ISSUE THREAD

**Issue by hhorak** [sclorg/rhscl2dockerfile#19]
I've kind of expected some discussion already happening at this point about sti scripts, but since I'm not seeing it anywhere, I'd like open this topic here (let me know if it was already consulted somewhere).

The sti scripts in openshift are available here:
* https://github.com/openshift/sti-php/tree/master/5.5/.sti/bin
* https://github.com/openshift/sti-python/tree/master/3.3/.sti/bin
* https://github.com/openshift/sti-ruby/tree/master/2.0/.sti/bin
* https://github.com/openshift/sti-nodejs/tree/master/0.10/.sti/bin

There are no comments in the scripts, so I'd like to first document here (and later as comments in the scripts):
* what is the purpose of these scripts
* who will run them (if they would be usable only in openshift, then I propose to prefix them with 'openshift-' name)

Also, I'd like to hear from RHSCL maintainers, if those scripts are something they may maintain in the RHSCL images.

@humaton @sYnfo @rkuska @strzibny @remicollet @mfojtik @jhadvig @bparees @soltysh

**Comment by mfojtik** [comment]
More documentation:

https://github.com/openshift/source-to-image/blob/master/docs/builder_image.md
https://github.com/openshift/source-to-image/blob/master/README.md
http://docs.openshift.org/latest/creating_images/sti.html

**Comment by mfojtik** [comment]
@hhorak the documentation above also answers your questions.

**Comment by hhorak** [comment]
Adding @ppisar for perl...
So, let's include at least the links to documentation above as comments in the sti-scripts.

**Comment by hhorak** [comment]
About including the sti scripts in the RHSCL images, it seems we found consensus with @mfojtik, so hopefully it will work for others as well -- those sti scripts don't have to be located in `$PATH`, so let's include them in `/usr/share/openshift/bin/assemble`, `/usr/share/openshift/bin/run` and `/usr/share/openshift/bin/usage` (this `usage` script is a different usage than what `docker run IMAGE container-usage` returns).

The '.../bin/...' part is important for some reason in openshift, but I don't see a problem to respect it under `/usr/share/openshift/`..

Another important thing is that we're required to set `STI_SCRIPTS_URL` environment variable to `/usr/share/openshift` -- @mfojtik, the `bin` part shouldn't be part of the env. variable, right?

The purpose of the sti scripts is that they will be used by openshift users only and ignored by non-openshift users. That said, the scripts simply should work, please, take a look if there is some issue and raise your hand asap if there is something we can't add into RHSCL.

Otherwise, please, add the scripts into the RHSCL images. Ping me for more info about how to add scripts into the generator :)

**Comment by soltysh** [comment]
>Another important thing is that we're required to set STI_SCRIPTS_URL environment variable to /usr/share/openshift -- @mfojtik, the bin part shouldn't be part of the env. variable, right?

`STI_SCRIPTS_URL` will have to be `/usr/share/openshift/bin/` you cannot ommit the `bin` part.

**Comment by strzibny** [comment]
I was looking on them briefly and the problems from Ruby POV I see are:

- only support for **puma** application server
- no support for db migration

which means it's not really a general use case for customers.

- dependency on JavaScript runtime
  - if they should be included in *rh-ruby22*, then they carry external dependency to the image

So we should probably make clear they are *somehow labeled* as OpenShift's.

**Comment by hhorak** [comment]
@strzibny wrote:
> - dependency on JavaScript runtime
  - if they should be included in *rh-ruby22*, then they carry external dependency to the image

By 'external dependency' you mean the nodejs collection we talked about or anything else?

> So we should probably make clear they are *somehow labeled* as OpenShift's.

You mean the sti scripts, right? I think since the path they are located under includes openshift, it should serve as kind of a label. It should be also possible to have a README in the same directory that would include the links above and would state that these are scripts usable mainly in openshift.

**Comment by mfojtik** [comment]
@strzibny puma is default because we have to pick one which is reasonable... migrations are to be solved by DeploymentConfig Post-Exec hooks as that is not trivial problem to solve. The nodejs dependency is there to support Rails use-cases.

**Comment by hhorak** [comment]
@mfojtik wrote:
> The nodejs dependency is there to support Rails use-cases.

openshift will use the `rh-rails41` image anyway, where nodejs is already installed/enabled, right?

**Comment by jhadvig** [comment]
@hhorak are you going to create separate images for ruby apps and rails apps? Dont think thats the right approache. If so, all ruby apps will need to be based on that rails image, not just rails apps, since the bundler is only part of the rails scl.

**Comment by hhorak** [comment]
@jhadvig that is no more true for rh-ruby22, bundler is part of rh-ruby22 actually, rh-ror41 then extends the rh-ruby22... that's why ruby without rails image is quite well usable but I expect you'll use rh-ror41 anyway in openshift, right?

**Comment by strzibny** [comment]
@jhadvig @hhorak Yeah, thats why I mentioned it, because you can now use just `rh-ruby22`, but you would have to link a nodejs container to use the sti scripts..

@mfojtik I did not say that Puma is not a reasonable choice, but we actually have Passenger packages within RHSCL. So I only meant that for users _outside_ of OpenShift sti provide only one option and it's not even one that would consume our RHSCL content.

**Comment by jhadvig** [comment]
@hhorak good to know about that change. About the usage of the ror image, Im not sure if we will use only that image cause for sinatra apps its pointless to have most of the ror packages(activerecord, ... ).

Also I didn't get what @strzibny meant with with linking nodejs container for usage of the sti scripts? That doesn't seem to be the right way to do, cause we need to launch the sti scripts inside the application container and not starting additional container, in which the scripts will run (can even imagine how this flow could work without having to do some additional hacks). Does that mean that the ruby and ror images wont have nodejs + v8 scl enabled? If so, how will the asset compilation be working  for the ror apps(dont forget that the compilation of assets could be done also in regular ruby application not just ror) ? Or it will also take place in the linked nodejs container?  
All this doesn't seems right and for me is a very hackish approach compared the 'wheel' we already created.

**Comment by strzibny** [comment]
@jhadvig My point was, that only `rh-ror41` will have nodejs inside and enabled. At least that's what I understood.

**Comment by jhadvig** [comment]
@strzibny right but as I mention, I you are going to deliver to set of ruby images, both of them will have to have nodejs inside and enable, bacause of the asset compilation, which could be done also in regular ruby application(its just a gem which is doing all the magic)

**Comment by strzibny** [comment]
@jhadvig I thought we did say that rh-ruby22 will be on it's own, _if not_, then my complain about sti having an external dependency is indeed invalid.

**Comment by mfojtik** [comment]
@hhorak is there a reason to move the scripts to `/usr/share/openshift/bin/`?
The tool name is 'STI' not OpenShift. STI scripts are not 'OpenShift' specific, the STI tool is standalone and works on its own.

**Comment by mfojtik** [comment]
@jhadvig we discussed Ruby while you were on PTO, the resolution was that we will ship:

* ruby that will be only ruby + bundler (no nodejs)
* rails that will contain the above + rails + nodejs

You don't need nodejs as long as you don't have Rails with assets. @strzibny please correct me if I'm wrong, it was a long time ago and my memory is getting old :-)

**Comment by mfojtik** [comment]
@strzibny we were evaluating passenger in OpenShift (we currently use it in OpenShift v2) and there are **many** problems with passenger itself ;-) Some that comes to my mind:

* require httpd
* require specific version of rack to be installed in system (and then conflict with latest rails rack)... I can tell you stories about this problems in OpenShif v2 :)
* it is more complex to configure and maintain the configuration
* we want to make the # of processes in container minimal
* it provides 'load-balancing' internally by pre-forking passenger processes, which is useless in container as the containers will scale by kubernetes, so there is no need to scale it internally...

These are just some of the problem we had with passenger. We choose 'puma' because it is single process (can be multiple if you want). It has good performance and it works with jruby which will provide future compatibility. It also does not require any configuration mechanism.

Does the above make sense to you?

**Comment by strzibny** [comment]
@mfojtik Yes, you are right and here https://github.com/openshift/sti-ruby/blob/master/2.0/.sti/bin/assemble#L7 we check explicitly for Rails. (But that does not mean that a Sinatra or Cuba app wouldn't need to compile assets. It would depend on the application.)

And thanks for the clarification on Passenger.

**Comment by mfojtik** [comment]
@strzibny yes, but Sinatra or Cuba does not fit into 80% of use-cases we want to cover ;-) We can't satisfy everybody, if you require something extra, then you will have to perform extra work to setup it, like write your own assemble script or put `.sti/bin/assemble` into your application source repository.

**Comment by jhadvig** [comment]
@mfojtik @strzibny I see your point, thanks for enlightenment :+1:

**Comment by hhorak** [comment]
@mfojtik wrote:
> @hhorak is there a reason to move the scripts to /usr/share/openshift/bin/?
The tool name is 'STI' not OpenShift. STI scripts are not 'OpenShift' specific, the STI tool is standalone and works on its own.

No big reason, I'm fine with `/usr/share/sti/bin` as well if you are..

**Comment by mfojtik** [comment]
@hhorak +1

**Comment by jhadvig** [comment]
@hhorak +1

**Comment by rkuska** [comment]
Python's STI scrips looks good to me and also to @sYnfo 

Just a note (bikeshedding mostly), from the usage bin file:

sti build https://github.com/openshift/sti-python.git --contextDir=3.3/test/setup-test-app/ openshift/python-33-${DISTRO}7 python-sample-app

it's a bit confusing using 3.3/33 in a path when we will ship a python34.

**Comment by soltysh** [comment]
@rkuska yeah, we can parametrize that part as well, to make it match current produced image.

**Comment by hhorak** [comment]
@rkuska wrote:
> it's a bit confusing using 3.3/33 in a path when we will ship a python34.

Yes, in RHSCL the paths will be slightly different, I think that is expected and fine.

**Comment by soltysh** [comment]
@rkuska when we'll be shipping python 3.4, we'll create additional directory and name the image appropriately. Currently we're using 3.3, because this is the latest available from SCL (1.2).

**Comment by rkuska** [comment]
FYI @synfo added a python-pip [0] to python27 (and python33) docker file (it is installed by default only in rh-python34) as it is used by STI files (also STI files are present [1]).

[0] 64d88b1ed499862aa4f3250d730e5b351101856c
[1] 0a7fac5661101822ed26dffce817d2b5524a1571

**Comment by ppisar** [comment]
Regarding Perl:

The dockerfile installs third-party code from the internet (Dockerfile.rhel7: scl enable perl516 -- cpanp 'i App::cpanminus'). This is not acceptable for supported images and it will probably not work in restricted build environment. One could move this step into .sti/bin/assemble phase. Or one could reimplement the functionality (install CPAN modules listed in a cpanfile) or bundle the App-cpanminus by ADD directive.

The dockerfile as well as .sti/bin/run script moves httpd files to non-default path (/opt/openshift). Without tuning selinux file labeling, it will cause SELinux denials preventing httpd from accessing the files. Either you should use the default directories, or amend SELinux file contexts (semanage fcontext --add && restorecon -FR), or disable SELinux. (I do not recomend the last option, I prefer the first one).

The .sti/bin/assemble defines some arbitrary paths for installing Perl modules from CPAN to (extlib/lib/perl5 and extlib). While it does not harm anything, I cannot see reason why to do that. Provided the image is built as root, one could keep installing the modules into standard Perl paths (/opt/rh/perl516/root/usr/local/{lib64,share}/perl5) by removing all the -l and -I options.

**Comment by strzibny** [comment]
@mfojtik @jhadvig Just to be clear that we are on the same boat:

- we ship `rh-ruby22` container without Node.js and *without* sti scripts.
- we ship `rh-ror41` with `rh-ruby22`, Node.js and sti scripts *included*

Thanks to that we will have  both clean Ruby container and Rails container that will fit OpenShift usecases and provides sti scripts for a straightforward build of applications. Is that correct? Or do you still see a reason for inclusion of sti scripts in `rh-ruby22` container?

**Comment by mfojtik** [comment]
@strzibny we are not doing that for this release (we will ship only sti-ruby as it is now) as the new image will have to pass through our QA. But this is something we want to do long term and applies to other frameworks.

**Comment by mnagy** [comment]
@ppisar

It seems to me that reimplementing the functionality to install CPAN modules listed in a cpanfile would be the best solution, however, we don't have the know-how for that. Packaging cpanm to the perl collection, but I believe that would take too long?

Good catch with SELinux. I think this should be fixed (for now) by amending SELinux contexts in docker build. I've opened an issue here: https://github.com/openshift/sti-perl/issues/14

With regards to paths, we prefer to build as non-root.

**Comment by hhorak** [comment]
@mnagy wrote:
> Packaging cpanm to the perl collection, but I believe that would take too long?

Yes, it is unfortunately too late now for rhscl 2.0.

**Comment by mfojtik** [comment]
@hhorak is there a way to sneak that module into the image (like importing it to dist-git?) if not and we will not solve this in time I would vote to move the cpanm installation into assemble script and install it during the application build time. That way you guys can build the image and we can keep the installation workflow we have now...

@danmcp ^^

**Comment by hhorak** [comment]
@mfojtik wrote:
> @strzibny we are not doing that for this release (we will ship only sti-ruby as it is now) as the new image will have to pass through our QA. But this is something we want to do long term and applies to other frameworks.

I have problem decoding if that meant ack for what @strzibny proposed or something different :)

**Comment by mfojtik** [comment]
@hhorak long term ack, short term (3.0 release) we will ship just 'ruby' image with all collections enabled (it is not trivial to add another image to the loop for us at this time)

**Comment by hhorak** [comment]
@mfojtik wrote:
> @hhorak is there a way to sneak that module into the image (like importing it to dist-git?) if not and we will not solve this in time I would vote to move the cpanm installation into assemble script and install it during the application build time. That way you guys can build the image and we can keep the installation workflow we have now...

-1 for sneaking the module into image, it is way too hacky; moving the installation into the assemble script seems much better to me

**Comment by ppisar** [comment]
On Tue, May 26, 2015 at 07:01:15AM -0700, Michal Fojtik wrote:
> @hhorak is there a way to sneak that module into the image (like importing
> it to dist-git?)

The App-cpanminus is one large bundle. (See
<https://bugzilla.redhat.com/show_bug.cgi?id=1045378>). Latest version would
missing these modules in perl516 collection after unbundling:

CPAN::DistnameInfo
CPAN::Meta::Check
File::pushd
Module::CPANfile
Parse::PMFile
String::ShellQuote

I can try to reimplemnt the functionality using missing Module::CPANfile
module on top of existing `CPAN' module, but I'm not sure if it makes sense as
interim solution as it would change behaviour very probably.

-- Petr

**Comment by hhorak** [comment]
@ppisar wrote:
> I can try to reimplemnt the functionality using missing Module::CPANfile
module on top of existing `CPAN' module, but I'm not sure if it makes sense as
interim solution as it would change behaviour very probably.

I don't think this is worth the trouble, if we can use the assemble script way.

**Comment by mnagy** [comment]
+1 for the assemble script. Who will do it?

**Comment by mfojtik** [comment]
@mnagy create issue for sti-perl and explain why we have to do that

**Comment by mfojtik** [comment]
@hhorak is there a way to include the missing package in next version of RHCSL? does it require to fill some BZ or some process that we can help with?

**Comment by hhorak** [comment]
@mfojtik Yes, please fill the BZ and talk to our PM as well, so it get's appropriate attention ;)

**Comment by mnagy** [comment]
@ppisar I've looked at the perl image and SELinux labels are the same.

```
-rw-rw-r--. default default system_u:object_r:svirt_sandbox_file_t:s0:c392,c985 /opt/openshift/httpd.conf
-rw-r--r--. root    root    system_u:object_r:svirt_sandbox_file_t:s0:c392,c985 /opt/rh/httpd24/root/etc/httpd/conf/httpd.conf
```

Am I missing something? Or are the SELinux labels in the SCL package wrong as well?

**Comment by ppisar** [comment]
On Mon, Jun 08, 2015 at 04:26:19AM -0700, Martin Nagy wrote:
> @ppisar I've looked at the perl image and SELinux labels are the same.
> 
> ```
> -rw-rw-r--. default default system_u:object_r:svirt_sandbox_file_t:s0:c392,c985 /opt/openshift/httpd.conf
> -rw-r--r--. root    root    system_u:object_r:svirt_sandbox_file_t:s0:c392,c985 /opt/rh/httpd24/root/etc/httpd/conf/httpd.conf
> ```
> 
> Am I missing something? Or are the SELinux labels in the SCL package wrong as well?
> 
The label is not httpd_config_t.

Obviously now all the processes and files are squashed to the
svirt_sandbox_file_t label. While this prevents container processes to
interact with objects out of container, it does not build ony wall inside the
container.

I don't know if this the desired state of SELinux application on containers.
I was assuming that it was a bug. Providing there is only one user in the
container it turns the whole security model into the era of DOS.

-- Petr

**Comment by mnagy** [comment]
@rhatdan can you please advice? https://github.com/sclorg/rhscl2dockerfile/issues/19#issuecomment-109979514
